protocol/testing/e2e/funding/funding_e2e_test.go

```diff
@@ -0,0 +1,247 @@
+	epochstypes "github.com/dydxprotocol/v4-chain/protocol/x/epochs/types"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	"github.com/stretchr/testify/require"
+)
+
+type TestHumanOrder struct {
+	Order      clobtypes.Order
+	HumanPrice string
+	HumanSize  string
+}
+
+const (
+	BlockTimeDuration             = 2 * time.Second
+	NumBlocksPerMinute            = int64(time.Minute / BlockTimeDuration) // 30
+	BlockHeightAtFirstFundingTick = 1000
+)
+
+var (
+	OrderTemplate_Alice_Num0_Id0_Clob0_Buy_LongTerm = clobtypes.Order{
+		OrderId: clobtypes.OrderId{
+			SubaccountId: constants.Alice_Num0,
+			ClientId:     0,
+			OrderFlags:   clobtypes.OrderIdFlags_LongTerm,
+			ClobPairId:   0,
+		},
+		Side: clobtypes.Order_SIDE_BUY,
+		GoodTilOneof: &clobtypes.Order_GoodTilBlockTime{
+			GoodTilBlockTime: uint32(GenesisTime.Add(24 * time.Hour).Unix()),
+		},
+	}
+	OrderTemplate_Alice_Num0_Id1_Clob0_Buy_LongTerm = clobtypes.Order{
+		OrderId: clobtypes.OrderId{
+			SubaccountId: constants.Alice_Num0,
+			ClientId:     1,
+			OrderFlags:   clobtypes.OrderIdFlags_LongTerm,
+			ClobPairId:   0,
+		},
+		Side: clobtypes.Order_SIDE_BUY,
+		GoodTilOneof: &clobtypes.Order_GoodTilBlockTime{
+			GoodTilBlockTime: uint32(GenesisTime.Add(24 * time.Hour).Unix()),
+		},
+	}
+	OrderTemplate_Bob_Num0_Id0_Clob0_Sell_LongTerm = clobtypes.Order{
+		OrderId: clobtypes.OrderId{
+			SubaccountId: constants.Bob_Num0,
+			ClientId:     0,
+			OrderFlags:   clobtypes.OrderIdFlags_LongTerm,
+			ClobPairId:   0,
+		},
+		Side: clobtypes.Order_SIDE_SELL,
+		GoodTilOneof: &clobtypes.Order_GoodTilBlockTime{
+			GoodTilBlockTime: uint32(GenesisTime.Add(24 * time.Hour).Unix()),
+		},
+	}
+	OrderTemplate_Bob_Num0_Id1_Clob0_Sell_LongTerm = clobtypes.Order{
+		OrderId: clobtypes.OrderId{
+			SubaccountId: constants.Bob_Num0,
+			ClientId:     1,
+			OrderFlags:   clobtypes.OrderIdFlags_LongTerm,
+			ClobPairId:   0,
+		},
+		Side: clobtypes.Order_SIDE_SELL,
+		GoodTilOneof: &clobtypes.Order_GoodTilBlockTime{
+			GoodTilBlockTime: uint32(GenesisTime.Add(24 * time.Hour).Unix()),
+		},
+	}
+	// Genesis time of the chain
+	GenesisTime                               = time.Unix(1690000000, 0)
+	FirstFundingSampleTick                    = time.Unix(1690000050, 0)
+	FirstFundingTick                          = time.Unix(1690002000, 0)
+	LastFundingSampleOfSecondFundingTickEpoch = time.Unix(1690005570, 0)
+	SecondFundingTick                         = time.Unix(
+		1690002000+int64(epochstypes.FundingTickEpochDuration),
+		0,
+	)
+)
+
+func TestFunding(t *testing.T) {
+	tests := map[string]struct {
+		testHumanOrders   []TestHumanOrder
+		initialIndexPrice map[uint32]string
+		// index price to be used in premium calculation
+		indexPriceForPremium map[uint32]string
+		// oracle price for funding index calculation
+		oracelPriceForFundingIndex map[uint32]string
+		expectedFundingPremium     int32
+		expectedFundingIndex       int64
+	}{
+		"Test funding": {
```

Comment:
should we in the future add cases where funding is 0?

Reply:
Yes, planning to do that in a follow-up PR. Added a TODO.